File: samples/grpc/local-drone-control-scala/src/main/scala/charging/ChargingStation.scala
package charging

import akka.Done
import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.ActorContext
import akka.actor.typed.scaladsl.Behaviors
import akka.pattern.StatusReply
import akka.persistence.typed.ReplicaId
import akka.persistence.typed.scaladsl.Effect
import akka.persistence.typed.scaladsl.EventSourcedBehavior
import akka.persistence.typed.scaladsl.ReplicationContext
import akka.projection.grpc.consumer.ConsumerFilter
import akka.projection.grpc.replication.scaladsl.ReplicatedBehaviors
import akka.projection.grpc.replication.scaladsl.Replication
import akka.projection.grpc.replication.scaladsl.Replication.EdgeReplication
import akka.projection.grpc.replication.scaladsl.ReplicationSettings
import akka.projection.r2dbc.scaladsl.R2dbcReplication
import akka.serialization.jackson.CborSerializable

import java.time.Instant
import scala.concurrent.duration._

object ChargingStation {

  // #commands
  sealed trait Command extends CborSerializable
  case class Create(
      locationId: String,
      chargingSlots: Int,
      replyTo: ActorRef[StatusReply[Done]])
      extends Command
  case class StartCharging(
      droneId: String,
      replyTo: ActorRef[StatusReply[StartChargingResponse]])
      extends Command
  sealed trait StartChargingResponse extends CborSerializable
  case class AllSlotsBusy(firstSlotFreeAt: Instant)
      extends StartChargingResponse

  case class GetState(replyTo: ActorRef[StatusReply[ChargingStation.State]])
      extends Command

  case class CompleteCharging(
      droneId: String,
      reply: ActorRef[StatusReply[Done]])
      extends Command
  // #commands

  // #events
  sealed trait Event extends CborSerializable
  case class Created(locationId: String, chargingSlots: Int) extends Event
  case class ChargingStarted(droneId: String, expectedComplete: Instant)
      extends Event
      with StartChargingResponse

  case class ChargingCompleted(droneId: String) extends Event
  // #events

  // #state
  case class ChargingDrone(
      droneId: String,
      expectedComplete: Instant,
      replicaId: String)
  case class State(
      chargingSlots: Int,
      dronesCharging: Set[ChargingDrone],
      stationLocationId: String)
      extends CborSerializable
  // #state

  val EntityType = "charging-station"

  private val FullChargeTime = 5.minutes

  /**
   * Init for running in edge node, this is the only difference from the ChargingStation
   * in restaurant-deliveries-service
   */
  // #edgeReplicaInit
  def initEdge(locationId: String)(
      implicit system: ActorSystem[_]): EdgeReplication[Command] = {
    val replicationSettings =
      ReplicationSettings[Command](EntityType, R2dbcReplication())
        .withSelfReplicaId(ReplicaId(locationId))
        .withInitialConsumerFilter(
          // only replicate charging stations local to the edge system
          Seq(
            ConsumerFilter.excludeAll,
            ConsumerFilter.IncludeTopics(Set(locationId))))
    Replication.grpcEdgeReplication(replicationSettings)(ChargingStation.apply)
  }
  // #edgeReplicaInit

  /**
   * Init for running in cloud replica
   */
  // #replicaInit
  def init(implicit system: ActorSystem[_]): Replication[Command] = {
    val replicationSettings =
      ReplicationSettings[Command](EntityType, R2dbcReplication())
        .withEdgeReplication(true)
    Replication.grpcReplication(replicationSettings)(ChargingStation.apply)
  }
  // #replicaInit

  def apply(
      replicatedBehaviors: ReplicatedBehaviors[Command, Event, Option[State]])
      : Behavior[Command] = {
    Behaviors.setup[Command] { context =>
      replicatedBehaviors.setup { replicationContext =>
        context.log
          .info("Charging Station {} starting up", replicationContext.entityId)
        new ChargingStation(context, replicationContext).behavior()
      }
    }
  }

}

class ChargingStation(
    context: ActorContext[ChargingStation.Command],
    replicationContext: ReplicationContext) {

  import ChargingStation._

  def behavior(): EventSourcedBehavior[Command, Event, Option[State]] =
    EventSourcedBehavior(
      replicationContext.persistenceId,
      None,
      handleCommand,
      handleEvent)
      // tag with location id so we can replicate only to the right edge node
      // #tagging
      .withTaggerForState {
        case (None, _)        => Set.empty
        case (Some(state), _) => Set("t:" + state.stationLocationId)
      }
  // #tagging

  // #commandHandler
  private def handleCommand(
      state: Option[State],
      command: Command): Effect[Event, Option[State]] =
    state match {
      case None        => handleCommandNoState(command)
      case Some(state) => handleCommandInitialized(state, command)
    }

  private def handleCommandNoState(
      command: Command): Effect[Event, Option[State]] =
    command match {
      case Create(locationId, chargingSlots, replyTo) =>
        Effect
          .persist(Created(locationId, chargingSlots))
          .thenReply(replyTo)(_ => StatusReply.Ack)
      case StartCharging(_, replyTo) =>
        Effect.reply(replyTo)(
          StatusReply.Error(
            s"Charging station ${replicationContext.entityId} not initialized"))
      case GetState(replyTo) =>
        Effect.reply(replyTo)(
          StatusReply.Error(
            s"Charging station ${replicationContext.entityId} not initialized"))
      case unexpected =>
        context.log.warn(
          "Got an unexpected command {} but charging station with id {} not initialized",
          unexpected.getClass,
          replicationContext.entityId)
        Effect.none
    }

  private def handleCommandInitialized(
      state: ChargingStation.State,
      command: ChargingStation.Command): Effect[Event, Option[State]] = {
    command match {
      case Create(_, _, replyTo) =>
        Effect.reply(replyTo)(StatusReply.Error(
          s"Got a create command, but station id ${replicationContext.entityId} was already created"))

      case StartCharging(droneId, replyTo) =>
        if (state.dronesCharging.exists(_.droneId == droneId)) {
          context.log.warn(
            "Drone {} requested charging but is already charging. Ignoring.",
            droneId)
          Effect.none
        } else if (state.dronesCharging.size >= state.chargingSlots) {
          val earliestFreeSlot =
            state.dronesCharging.map(_.expectedComplete).min
          context.log.info(
            "Drone {} requested charging but all stations busy, earliest free slot {}",
            droneId,
            earliestFreeSlot)
          Effect.reply(replyTo)(
            StatusReply.Success(AllSlotsBusy(earliestFreeSlot)))
        } else {
          // charge
          val expectedComplete =
            Instant.now().plusSeconds(FullChargeTime.toSeconds)
          context.log.info(
            "Drone {} requested charging, expected to complete charging at {}",
            droneId,
            expectedComplete)
          val event = ChargingStarted(droneId, expectedComplete)
          Effect
            .persist(event)
            .thenReply(replyTo)(_ => StatusReply.Success(event))
        }

      case CompleteCharging(droneId, replyTo) =>
        if (state.dronesCharging.exists(_.droneId == droneId)) {
          context.log.info("Drone {} completed charging", droneId)
          Effect
            .persist(ChargingCompleted(droneId))
            .thenReply(replyTo)(_ => StatusReply.Ack)
        } else {
          Effect.reply(replyTo)(
            StatusReply.error(s"Drone $droneId is not currently charging"))
        }

      case GetState(replyTo) =>
        Effect.reply(replyTo)(StatusReply.Success(state))

    }
  }
  // #commandHandler

  // #eventHandler
  private def handleEvent(state: Option[State], event: Event): Option[State] = {
    state match {
      case None =>
        event match {
          case Created(locationId, chargingSlots) =>
            Some(State(chargingSlots, Set.empty, locationId))
          case unexpected =>
            throw new IllegalArgumentException(
              s"Got unexpected event ${unexpected} for uninitialized state")
        }

      case Some(state) =>
        event match {
          case Created(_, _) =>
            context.log.warn("Saw a second created event, ignoring")
            Some(state)
          case ChargingStarted(droneId, expectedComplete) =>
            Some(
              state.copy(dronesCharging = state.dronesCharging + ChargingDrone(
                droneId,
                expectedComplete,
                replicationContext.origin.id)))
          case ChargingCompleted(droneId) =>
            Some(
              state.copy(dronesCharging =
                state.dronesCharging.filterNot(_.droneId == droneId)))
        }

    }
  }
  // #eventHandler

}

File: samples/grpc/local-drone-control-java/src/main/java/charging/ChargingStation.java
package charging;

import akka.Done;
import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.pattern.StatusReply;
import akka.persistence.typed.ReplicaId;
import akka.persistence.typed.javadsl.*;
import akka.projection.grpc.consumer.ConsumerFilter;
import akka.projection.grpc.replication.javadsl.EdgeReplication;
import akka.projection.grpc.replication.javadsl.ReplicatedBehaviors;
import akka.projection.grpc.replication.javadsl.Replication;
import akka.projection.grpc.replication.javadsl.ReplicationSettings;
import akka.projection.r2dbc.javadsl.R2dbcReplication;
import akka.serialization.jackson.CborSerializable;
import com.fasterxml.jackson.annotation.JsonCreator;
import java.time.Duration;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

public class ChargingStation
    extends ReplicatedEventSourcedBehavior<
        ChargingStation.Command, ChargingStation.Event, ChargingStation.State> {

  // #commands
  public interface Command extends CborSerializable {}

  public static final class Create implements Command {
    public final String locationId;
    public final int chargingSlots;
    public final ActorRef<StatusReply<Done>> replyTo;

    public Create(String locationId, int chargingSlots, ActorRef<StatusReply<Done>> replyTo) {
      this.locationId = locationId;
      this.chargingSlots = chargingSlots;
      this.replyTo = replyTo;
    }
  }

  public static final class StartCharging implements Command {
    public final String droneId;
    public final ActorRef<StatusReply<StartChargingResponse>> replyTo;

    public StartCharging(String droneId, ActorRef<StatusReply<StartChargingResponse>> replyTo) {
      this.droneId = droneId;
      this.replyTo = replyTo;
    }
  }

  public interface StartChargingResponse extends CborSerializable {}

  public static final class AllSlotsBusy implements StartChargingResponse {
    public final Instant firstSlotFreeAt;

    @JsonCreator
    public AllSlotsBusy(Instant firstSlotFreeAt) {
      this.firstSlotFreeAt = firstSlotFreeAt;
    }
  }

  public static final class GetState implements Command {
    public final ActorRef<StatusReply<State>> replyTo;

    @JsonCreator
    public GetState(ActorRef<StatusReply<State>> replyTo) {
      this.replyTo = replyTo;
    }
  }

  public static final class CompleteCharging implements Command {
    final String droneId;

    final ActorRef<StatusReply<Done>> replyTo;

    public CompleteCharging(String droneId, ActorRef<StatusReply<Done>> replyTo) {
      this.droneId = droneId;
      this.replyTo = replyTo;
    }
  }
  // #commands

  // #events
  public interface Event extends CborSerializable {}

  public static final class Created implements Event {
    public final String locationId;
    public final int chargingSlots;

    public Created(String locationId, int chargingSlots) {
      this.locationId = locationId;
      this.chargingSlots = chargingSlots;
    }
  }

  public static final class ChargingStarted implements Event, StartChargingResponse {
    public final String droneId;
    public final Instant expectedComplete;

    public ChargingStarted(String droneId, Instant expectedComplete) {
      this.droneId = droneId;
      this.expectedComplete = expectedComplete;
    }
  }

  public static final class ChargingCompleted implements Event {
    public final String droneId;

    @JsonCreator
    public ChargingCompleted(String droneId) {
      this.droneId = droneId;
    }
  }

  public static final class ChargingDrone {
    public final String droneId;
    public final Instant expectedComplete;
    public final String replicaId;

    public ChargingDrone(String droneId, Instant expectedComplete, String replicaId) {
      this.droneId = droneId;
      this.expectedComplete = expectedComplete;
      this.replicaId = replicaId;
    }
  }
  // #events

  // #state
  public static final class State implements CborSerializable {
    public final int chargingSlots;
    public final Set<ChargingDrone> dronesCharging;
    public final String stationLocationId;

    public State(int chargingSlots, Set<ChargingDrone> dronesCharging, String stationLocationId) {
      this.chargingSlots = chargingSlots;
      this.dronesCharging = dronesCharging;
      this.stationLocationId = stationLocationId;
    }
  }
  // #state

  public static final String ENTITY_TYPE = "charging-station";

  private static final Duration FULL_CHARGE_TIME = Duration.ofMinutes(5);

  /**
   * Init for running in edge node, this is the only difference from the ChargingStation in
   * restaurant-deliveries-service
   */
  // #edgeReplicaInit
  public static EdgeReplication<Command> initEdge(ActorSystem<?> system, String locationId) {
    var replicationSettings =
        ReplicationSettings.create(
                Command.class, ENTITY_TYPE, R2dbcReplication.create(system), system)
            .withSelfReplicaId(new ReplicaId(locationId))
            .withInitialConsumerFilter(
                List.of(
                    // only replicate charging stations local to the edge system
                    ConsumerFilter.excludeAll(),
                    new ConsumerFilter.IncludeTopics(Set.of(locationId))));
    return Replication.grpcEdgeReplication(replicationSettings, ChargingStation::create, system);
  }
  // #edgeReplicaInit

  /** Init for running in cloud replica */
  // #replicaInit
  public static Replication<Command> init(ActorSystem<?> system) {
    var replicationSettings =
        ReplicationSettings.create(
                Command.class, ENTITY_TYPE, R2dbcReplication.create(system), system)
            // FIXME remove once release out with flag in config (1.5.1-M2/GA)
            .withEdgeReplication(true);
    return Replication.grpcReplication(replicationSettings, ChargingStation::create, system);
  }
  // #replicaInit

  public static Behavior<Command> create(
      ReplicatedBehaviors<Command, Event, State> replicatedBehaviors) {
    return Behaviors.setup(
        (ActorContext<Command> context) ->
            replicatedBehaviors.setup(
                replicationContext -> {
                  context
                      .getLog()
                      .info("Charging Station {} starting up", replicationContext.entityId());
                  return new ChargingStation(context, replicationContext);
                }));
  }

  private final ActorContext<Command> context;

  public ChargingStation(ActorContext<Command> context, ReplicationContext replicationContext) {
    super(replicationContext);
    this.context = context;
  }

  @Override
  public State emptyState() {
    return null;
  }

  // #commandHandler
  @Override
  public CommandHandler<Command, Event, State> commandHandler() {
    var noStateHandler =
        newCommandHandlerBuilder()
            .forNullState()
            .onCommand(
                Create.class,
                (state, create) ->
                    Effect()
                        .persist(new Created(create.locationId, create.chargingSlots))
                        .thenReply(create.replyTo, stateAfter -> StatusReply.ack()))
            .onCommand(
                StartCharging.class,
                startCharging ->
                    Effect()
                        .reply(
                            startCharging.replyTo,
                            StatusReply.error(
                                "Charging station "
                                    + getReplicationContext().entityId()
                                    + " not initialized")))
            .onCommand(
                GetState.class,
                getState ->
                    Effect()
                        .reply(
                            getState.replyTo,
                            StatusReply.error(
                                "Charging station "
                                    + getReplicationContext().entityId()
                                    + " not initialized")))
            .onCommand(
                command -> true,
                unexpected -> {
                  context
                      .getLog()
                      .warn(
                          "Got an unexpected command {} but charging station with id {} not initialized",
                          unexpected.getClass(),
                          getReplicationContext().entityId());
                  return Effect().none();
                });

    var initializedHandler =
        newCommandHandlerBuilder()
            .forNonNullState()
            .onCommand(
                Create.class,
                create ->
                    Effect()
                        .reply(
                            create.replyTo,
                            StatusReply.error(
                                "Got a create command, but station id "
                                    + getReplicationContext().entityId()
                                    + " was already created")))
            .onCommand(StartCharging.class, this::handleStartCharging)
            .onCommand(CompleteCharging.class, this::handleCompleteCharging)
            .onCommand(
                GetState.class,
                (state, getState) -> Effect().reply(getState.replyTo, StatusReply.success(state)));

    return noStateHandler.orElse(initializedHandler).build();
  }

  private Effect<Event, State> handleStartCharging(State state, StartCharging startCharging) {
    if (state.dronesCharging.stream()
        .anyMatch(charging -> charging.droneId.equals(startCharging.droneId))) {
      return Effect().reply(startCharging.replyTo, StatusReply.error("Drone already charging"));
    } else if (state.dronesCharging.size() >= state.chargingSlots) {
      var earliestFreeSlot =
          state.dronesCharging.stream()
              .min(Comparator.comparing(chargingDrone -> chargingDrone.expectedComplete))
              .get()
              .expectedComplete;
      context
          .getLog()
          .info(
              "Drone {} requested charging but all stations busy, earliest free slot {}",
              startCharging.droneId,
              earliestFreeSlot);
      return Effect()
          .reply(startCharging.replyTo, StatusReply.success(new AllSlotsBusy(earliestFreeSlot)));
    } else {
      // charge
      var expectedComplete = Instant.now().plus(FULL_CHARGE_TIME);
      context
          .getLog()
          .info(
              "Drone {} requested charging, will complete charging at {}",
              startCharging.droneId,
              expectedComplete);
      var event = new ChargingStarted(startCharging.droneId, expectedComplete);
      return Effect()
          .persist(event)
          .thenReply(startCharging.replyTo, newState -> StatusReply.success(event));
    }
  }

  private Effect<Event, State> handleCompleteCharging(
      State state, CompleteCharging completeCharging) {
    context.getLog().info("Drone {} completed charging", completeCharging.droneId);
    if (state.dronesCharging.stream()
        .anyMatch(chargingDrone -> chargingDrone.droneId.equals(completeCharging.droneId)))
      return Effect()
          .persist(new ChargingCompleted(completeCharging.droneId))
          .thenReply(completeCharging.replyTo, newState -> StatusReply.ack());
    else
      return Effect()
          .reply(
              completeCharging.replyTo,
              StatusReply.error(
                  "Drone " + completeCharging.droneId + " is not currently charging."));
  }
  // #commandHandler

  @Override
  public EventHandler<State, Event> eventHandler() {
    var noStateHandler =
        newEventHandlerBuilder()
            .forNullState()
            .onEvent(
                Created.class,
                created ->
                    new State(created.chargingSlots, Collections.emptySet(), created.locationId));

    var initializedStateHandler =
        newEventHandlerBuilder()
            .forNonNullState()
            .onEvent(
                Created.class,
                (state, event) -> {
                  context.getLog().warn("Saw a second created event, ignoring");
                  return state;
                })
            .onEvent(
                ChargingStarted.class,
                (state, event) -> {
                  var newSet = new HashSet<>(state.dronesCharging);
                  newSet.add(
                      new ChargingDrone(
                          event.droneId,
                          event.expectedComplete,
                          getReplicationContext().origin().id()));
                  return new State(
                      state.chargingSlots,
                      Collections.unmodifiableSet(newSet),
                      state.stationLocationId);
                })
            .onEvent(
                ChargingCompleted.class,
                (state, event) -> {
                  var newSet =
                      state.dronesCharging.stream()
                          .filter(charging -> !charging.droneId.equals(event.droneId))
                          .collect(Collectors.toSet());
                  return new State(
                      state.chargingSlots,
                      Collections.unmodifiableSet(newSet),
                      state.stationLocationId);
                });

    return noStateHandler.orElse(initializedStateHandler).build();
  }

  // #tagging
  @Override
  public Set<String> tagsFor(State state, Event event) {
    if (state == null) return Set.of();
    else return Set.of("t:" + state.stationLocationId);
  }
  // #tagging
}
